pkg/retagger/job_single.go

```diff
@@ -0,0 +1,108 @@
+package retagger
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+)
+
+// SingleJob is a concrete job which can be executed.
+type SingleJob struct {
+	Source Source
+
+	Destination Destination
+
+	Options JobOptions
+}
+
+// Compile wraps this job in an array to keep consistency with the CompilableJob interface.
+func (job *SingleJob) Compile(r *Retagger) ([]SingleJob, error) {
+	return []SingleJob{*job}, nil
+}
+
+// Describe returns a string containing basic information about the job.
+func (job *SingleJob) Describe() string {
+	return fmt.Sprintf("%s:%s will be tagged as %s:%s with digest %s",
+		job.Source.Image, job.Source.Tag, job.Destination.Image, job.Destination.Tag, job.Source.SHA)
+}
+
+// Execute runs the job using the given Retagger instance.
+func (job *SingleJob) Execute(r *Retagger) error {
+	return r.executeSingleJob(*job)
+}
+
+// ShouldRetag examines the state of the job and the given Retagger's registry and returns whether the job should be run.
+func (job *SingleJob) ShouldRetag(r *Retagger) (bool, error) {
+
+	tagExists, err := r.registry.CheckImageTagExists(job.Destination.Image, job.Destination.Tag)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	return !tagExists, nil
+}
+
+// SingleJobFromJobDefinition converts a JobDefinition into a SingleJob.
+func SingleJobFromJobDefinition(jobDef *JobDefinition, r *Retagger) *SingleJob {
+	job := &SingleJob{
+		Source: GetSourceForJob(jobDef, r),
+
+		Options: jobDef.Options,
+	}
+	job.Destination = GetDestinationForJob(job, r)
+	return job
+}
+
+func getRepoHostForJob(j *JobDefinition, r *Retagger) string {
+	// Handle remote, Docker Hub, and Docker library image path formats.
+	registryPath, err := r.registry.GuessRegistryPath(j.SourceImage)
+	if err != nil {
+		return j.SourceImage // Fall back to trying to use given image name.
+	}
+	return registryPath.Hostname()
+}
+
+func getFullImageNameForJob(j *JobDefinition, r *Retagger) string {
+	registryPath, err := r.registry.GuessRegistryPath(j.SourceImage)
+	if err != nil {
+		return j.SourceImage // Fall back to trying to use given image name.
+	}
+	return strings.Trim(registryPath.Path, "/") // Remove leading slash.
+}
+
+// GetSourceForJob populates a Source object based on the given JobDefinition.
+func GetSourceForJob(jobDef *JobDefinition, r *Retagger) Source {
```

Comment:
The order in this file is all over the place. It's common in go to have public methods before private ones. We also usually follow the convention of having methods before functions. I would suggest the same here.

Reply:
Sorted